infrastructure/templates/public-api/parameters/main-dev.bicepparam

```diff
@@ -1,26 +1,15 @@
 using '../main.bicep'
 
-//Environment Params -------------------------------------------------------------------
-param domain = 'publicapi'
-param subscription = 's101'
+// Environment Params
+param subscription = 's101d01'
 param environmentName = 'Development'
 
-//Networking Params --------------------------------------------------------------------------
-param deploySubnets = true
+// PostgreSQL Database Params
+param postgreSqlAdminName = 'PostgreSQLAdmin'
+param postgreSqlAdminPassword = 'adminPassword'
```

Comment:
Think we should probably just get rid of these as defaults from all the `bicepparam` files to avoid accidentally deploying with insecure values.

Looking at the PR description, think you've potentially alluded to creating a random string as part of the deployment? Not sure this would be possible via the Bicep template as the deployments should be deterministic / idempotent: https://stackoverflow.com/questions/68063937/bicep-creating-random-string

Think the next best thing would be to manually generate the values beforehand and put them in our Key Vault for use in the template!

Reply:
I wish that I could get rid of the defaults!  I have tried and unfortunately I get hit by this annoying issue:

Given this command to deploy the resource group (this is the command run from the pipeline):

```
az deployment group create \
--resource-group dfe-development-dw \
--template-file /home/dwatson/programs/azure/build-agent/_work/1/s/infrastructure/templates/public-api/main.bicep \
--parameters /home/dwatson/programs/azure/build-agent/_work/1/s/infrastructure/templates/public-api/parameters/main-dev.bicepparam \
--parameters \
    resourceTags='{"CostCentre": "A Cost Centre", "CreatedBy": "WATSON, Duncan", "Department": "Data Directorate", "DeploymentRepoUrl": "https://github/blah2", "ServiceOwner": "A Service Owner", "Solution": "EES"}'
    postgreSqlAdminName='myuser' \
    postgreSqlAdminPassword='mypassword'
```

you can see that the parameters `postgreSqlAdminName` and `postgreSqlAdminPassword` are being supplied along with `resourceTags` any other parameters it can find within the `main-dev.bicepparam` file.

You'd think therefore that it would be valid for `postgreSqlAdminName` and `postgreSqlAdminPassword` to have no default values in `main.bicep` as they are being correctly supplied.  Surprisingly though, this results in:

```
ERROR: /home/dwatson/programs/azure/build-agent/_work/1/s/infrastructure/templates/public-api/parameters/main-dev.bicepparam(1,7) : Error BCP258: The following parameters are declared in the Bicep file but are missing an assignment in the params file: "postgreSqlAdminName", "postgreSqlAdminPassword".
```

So this message is basically saying that if there are params in `main.bicep` without default values, we *have* to include values for them in the `bicepparam` file provided, or else we get an error.

This is a pain to be honest.  I think it basically means that we need to provide dummy default values in `main.bicep` despite the fact that we're passing them in with every `az` command we're issue.

I'd be sorely tempted to get rid of the bicepparam files altogether in favour of just sourcing all environment-specific variables from environment-specific Variable Groups instead.  This way we don't run into this error and aren't forced to supply dummy default values.  This also prevents us from spreading our configuration across multiple sources - we'd know that all of our Infra config is within a set of Variable Groups.  The only downside is that we lose the advantages of  strongly-typed param usage.

In other news, the values for `postgreSqlAdminName` and `postgreSqlAdminPassword` as provided by the pipeline command above are actually sourced from Key Vault via a "EES Bicep Infrastructure secrets - dfe-development-dw" Variable Group that is linked to the appropriate Resource Group's Key Vault.  Note that you can't mix and match in Variable Groups between stuff from Key Vault and manually curated params, so we'd likely end up with at least Variable Group with open parameters in it and one Variable Group linked to Key Vault for each environment.

As for this issue, for now I've wired in Variable Groups and that's where the PSQL credentials are coming from, but I've also had to placate the Bicep gods by supplying default values too.

Reply:
Ouch, yep can practically feel your pain from the above!

Interesting that DevOps doesn't automatically `set -x`. If it makes you feel any better, pretty much every other CI provider (GitHub, GitLab and Circle) automatically sets this as sane default, so just feels like standard Microsoft user unfriendliness.

Looking at your deployment command example above, it looks like the line with `resourceTags` is missing the trailing `\` at the end of its line. Feels like this would explain why it wasn't working?

Can we double check this?

Reply:
Gads, I'm an idiot!  You're right - corrected with additional backslash now thanks :see_no_evil: